core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -827,159 +848,170 @@ public T next() {
             }
-            return true;
-        }
 
-        private boolean findPrevChunk() {
-            while (!nextChunkIter.hasNext()) {
-                ByteBuffer serializedMinKey = nextChunk.minKey;
-                Map.Entry e = skiplist.lowerEntry(serializedMinKey);
-                if (e == null) { // there no next chunk
-                    assert serializedMinKey == minKey;
-                    next = Chunk.NONE;
-                    nextHandle = null;
-                    return false;
-                }
-                nextChunk = (Chunk) e.getValue();
-                Chunk nextNext = nextChunk.next.getReference();
-                if (nextNext == null) {
-                    nextChunkIter = nextChunk.descendingIter(keySerializer.deserialize(serializedMinKey), false);
-                    continue;
-                }
-                ByteBuffer nextMinKey = nextNext.minKey;
-                if (comparator.compare(nextMinKey, serializedMinKey) < 0) {
-                    nextChunk = nextNext;
-                    nextMinKey = nextChunk.next.getReference().minKey;
-                    while (comparator.compare(nextMinKey, serializedMinKey) < 0) {
-                        nextChunk = nextChunk.next.getReference();
-                        nextMinKey = nextChunk.next.getReference().minKey;
-                    }
-                }
-                nextChunkIter = nextChunk.descendingIter(keySerializer.deserialize(serializedMinKey), false); // TODO check correctness
-            }
-            return true;
-        }
+            //Must get nextIndex before return to avoid nextIndex==0 after init
+            nextIndex = nextChunkIter.next();
 
-        private void descend() {
-            if (!findPrevChunk()) {
+            //
+            boolean res = lastKeyBeforeChunkDelete.compareAndSet(currentChunk, nextChunk);
+            if (res == false) {
+                // failed to swap, rebalance on currentChunk updated lastKeyBeforeChunkDelete
                 return;
             }
-            // now look for first key in range
-            next = nextChunkIter.next();
-            ByteBuffer key = getKey(next, nextChunk);
-
-            while (!inBounds(key)) {
-                if (tooLow(key)) {
-                    // if we reached a key that is too low then there is no key in range
-                    next = Chunk.NONE;
-                    nextHandle = null;
-                    return;
-                }
-                if (!findPrevChunk()) {
-                    return;
-                }
-                next = nextChunkIter.next();
-                key = getKey(next, nextChunk);
+
+            //Must sign iterator after prev stages
+            if (currentChunk != nextChunk) {
+                nextChunk.signIterator(this);
+            }
+
+            //Check if released for frozen and store last key
+            if (nextChunk.state() == Chunk.State.RELEASED) {
```

Comment:
shouldn't  we check this on every next() method?

Reply:
It is checked every next()